runtime/lua/vim/diagnostic.lua

```diff
@@ -2,22 +2,95 @@ local api, if_nil = vim.api, vim.F.if_nil
 
 local M = {}
 
----@enum DiagnosticSeverity
+--- @alias Diagnostic vim.Diagnostic
+--- @class vim.Diagnostic
+--- @field bufnr? integer
+--- @field lnum integer 0-indexed
+--- @field end_lnum? integer 0-indexed
+--- @field col integer 0-indexed
+--- @field end_col? integer 0-indexed
+--- @field severity? vim.diagnostic.Severity
+--- @field message string
+--- @field source? string
+--- @field code? string
+--- @field _tags? { deprecated: boolean, unnecessary: boolean}
+--- @field user_data? any arbitrary data plugins can add
+--- @field namespace? integer
+
+--- @class vim.diagnostic.Opts
+--- @field float? boolean|vim.diagnostic.Opts.Float
+--- @field update_in_insert? boolean
+--- @field underline? boolean|vim.diagnostic.Opts.Underline
+--- @field virtual_text? boolean|vim.diagnostic.Opts.VirtualText
+--- @field signs? boolean|vim.diagnostic.Opts.Signs
+--- @field severity_sort? boolean|{reverse?:boolean}
+
+--- @class vim.diagnostic.OptsResolved
+--- @field float vim.diagnostic.Opts.Float
+--- @field update_in_insert boolean
+--- @field underline vim.diagnostic.Opts.Underline
+--- @field virtual_text vim.diagnostic.Opts.VirtualText
+--- @field signs vim.diagnostic.Opts.Signs
+--- @field severity_sort {reverse?:boolean}
+
+--- @class vim.diagnostic.Opts.Float
```

Comment:
Please consider defining this `@class` annotation closer to `M.open_float` function (near L1505), since `M.open_float` owns this type.

Reply:
I wouldn't say `open_float()` owns this type. `vim.diagnostic.Opts` owns this just as much.